File: packages/provide-uterm-go/recording/recording_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package recording

import (
	"os"
	"path/filepath"
	"testing"
)

func intPtr(v int) *int { return &v }

func events(names ...string) []Event {
	out := make([]Event, 0, len(names))
	for i, name := range names {
		out = append(out, Event{"ts": float64(i), "event": name, "data": map[string]any{"i": float64(i)}})
	}
	return out
}

// storeLifecycle exercises the shared Store contract for file and memory
// stores.
func storeLifecycle(t *testing.T, s Store, wantPath bool) {
	t.Helper()
	const sid = "sess-1"

	if err := s.StartSession(sid, map[string]any{"who": "test"}); err != nil {
		t.Fatal(err)
	}
	if err := s.AppendEvents(sid, events("output", "input", "output")); err != nil {
		t.Fatal(err)
	}
	if err := s.EndSession(sid); err != nil {
		t.Fatal(err)
	}

	meta, err := s.RecordingMeta(sid)
	if err != nil {
		t.Fatal(err)
	}
	if !meta.Exists || meta.SizeBytes <= 0 || meta.SessionID != sid {
		t.Fatalf("meta = %+v", meta)
	}

	all, err := s.GetEntries(sid, Query{})
	if err != nil {
		t.Fatal(err)
	}
	// log_start + 3 events + log_stop
	if len(all) != 5 || all[0]["event"] != "log_start" || all[4]["event"] != "log_stop" {
		t.Fatalf("entries = %v", all)
	}
	if all[0]["data"].(map[string]any)["who"] != "test" {
		t.Fatalf("start data = %v", all[0])
	}

	// Event filter.
	outputs, err := s.GetEntries(sid, Query{Event: "output"})
	if err != nil || len(outputs) != 2 {
		t.Fatalf("outputs = %v (%v)", outputs, err)
	}

	// Offset pagination over filtered stream.
	page, err := s.GetEntries(sid, Query{Event: "output", Offset: intPtr(1), Limit: 5})
	if err != nil || len(page) != 1 {
		t.Fatalf("page = %v (%v)", page, err)
	}

	// Tail behavior: last N events.
	tail, err := s.GetEntries(sid, Query{Limit: 2})
	if err != nil || len(tail) != 2 || tail[1]["event"] != "log_stop" {
		t.Fatalf("tail = %v (%v)", tail, err)
	}

	// Limit clamping: limit > 500 behaves like 500; negative clamps to 1.
	if got, _ := s.GetEntries(sid, Query{Limit: 10_000}); len(got) != 5 {
		t.Fatalf("clamped = %v", got)
	}
	if got, _ := s.GetEntries(sid, Query{Limit: -3}); len(got) != 1 {
		t.Fatalf("min-clamped = %v", got)
	}

	// Offset past the end and negative offsets.
	if got, _ := s.GetEntries(sid, Query{Offset: intPtr(99)}); len(got) != 0 {
		t.Fatalf("past-end = %v", got)
	}
	if got, _ := s.GetEntries(sid, Query{Offset: intPtr(-5), Limit: 500}); len(got) != 5 {
		t.Fatalf("neg-offset = %v", got)
	}

	path, err := s.GetPath(sid)
	if err != nil {
		t.Fatal(err)
	}
	if wantPath && path == "" {
		t.Fatal("expected local path")
	}
	if !wantPath && path != "" {
		t.Fatalf("unexpected path %q", path)
	}

	// Unknown session: empty results, no error.
	meta, err = s.RecordingMeta("nope")
	if err != nil || meta.Exists {
		t.Fatalf("meta = %+v (%v)", meta, err)
	}
	if got, _ := s.GetEntries("nope", Query{}); len(got) != 0 {
		t.Fatalf("entries = %v", got)
	}
	if p, _ := s.GetPath("nope"); p != "" {
		t.Fatalf("path = %q", p)
	}
}

func TestLocalFileStoreLifecycle(t *testing.T) {
	storeLifecycle(t, NewLocalFileStore(t.TempDir()), true)
}

func TestInMemoryStoreLifecycle(t *testing.T) {
	storeLifecycle(t, NewInMemoryStore(), false)
}

func TestLocalFileStoreAppendWithoutStartReopens(t *testing.T) {
	dir := t.TempDir()
	s := NewLocalFileStore(dir)
	if err := s.AppendEvents("cold", events("output")); err != nil {
		t.Fatal(err)
	}
	got, err := s.GetEntries("cold", Query{})
	if err != nil || len(got) != 1 {
		t.Fatalf("got %v (%v)", got, err)
	}
	// EndSession without an open handle is a no-op.
	if err := s.EndSession("never-started-other"); err != nil {
		t.Fatal(err)
	}
}

func TestLocalFileStoreSkipsMalformedLines(t *testing.T) {
	dir := t.TempDir()
	s := NewLocalFileStore(dir)
	if err := s.AppendEvents("sess", events("output")); err != nil {
		t.Fatal(err)
	}
	// AppendEvents lazily opened a handle, so EndSession writes log_stop.
	if err := s.EndSession("sess"); err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(dir, "sess.jsonl")
	f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0o600)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := f.WriteString("{malformed\n"); err != nil {
		t.Fatal(err)
	}
	if err := f.Close(); err != nil {
		t.Fatal(err)
	}
	got, err := s.GetEntries("sess", Query{})
	if err != nil || len(got) != 2 {
		t.Fatalf("got %v (%v)", got, err)
	}
	// Offset variant also skips malformed lines.
	got, err = s.GetEntries("sess", Query{Offset: intPtr(0)})
	if err != nil || len(got) != 2 {
		t.Fatalf("got %v (%v)", got, err)
	}
	// Offset mode stops reading once the limit is filled.
	got, err = s.GetEntries("sess", Query{Offset: intPtr(0), Limit: 1})
	if err != nil || len(got) != 1 || got[0]["event"] != "output" {
		t.Fatalf("got %v (%v)", got, err)
	}
}

func TestLocalFileStoreStartErrors(t *testing.T) {
	dir := t.TempDir()
	blocker := filepath.Join(dir, "block")
	if err := os.WriteFile(blocker, nil, 0o600); err != nil {
		t.Fatal(err)
	}
	s := NewLocalFileStore(filepath.Join(blocker, "sub"))
	if err := s.StartSession("x", nil); err == nil {
		t.Fatal("expected error")
	}
	if err := s.AppendEvents("x", events("output")); err == nil {
		t.Fatal("expected error")
	}
}

func TestLocalFileStoreUnserializableEvent(t *testing.T) {
	s := NewLocalFileStore(t.TempDir())
	if err := s.AppendEvents("s", []Event{{"bad": make(chan int)}}); err == nil {
		t.Fatal("expected marshal error")
	}
}

func TestWriteEventsWriteError(t *testing.T) {
	// A read-only handle makes the write itself fail.
	path := filepath.Join(t.TempDir(), "ro.jsonl")
	if err := os.WriteFile(path, nil, 0o600); err != nil {
		t.Fatal(err)
	}
	f, err := os.Open(path)
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = f.Close() }()
	if err := writeEvents(f, events("output")); err == nil {
		t.Fatal("expected write error")
	}
}

func TestInMemoryStoreMetaSizeAndUnserializable(t *testing.T) {
	s := NewInMemoryStore()
	if err := s.AppendEvents("m", []Event{{"bad": make(chan int)}}); err != nil {
		t.Fatal(err)
	}
	if _, err := s.RecordingMeta("m"); err == nil {
		t.Fatal("expected marshal error in meta size")
	}
	// EndSession on a session never started still records log_stop without
	// flipping a session flag.
	if err := s.EndSession("m"); err != nil {
		t.Fatal(err)
	}
}

func TestNullStore(t *testing.T) {
	var s Store = NullStore{}
	if err := s.StartSession("x", nil); err != nil {
		t.Fatal(err)
	}
	if err := s.AppendEvents("x", events("output")); err != nil {
		t.Fatal(err)
	}
	if err := s.EndSession("x"); err != nil {
		t.Fatal(err)
	}
	meta, err := s.RecordingMeta("x")
	if err != nil || meta.Exists || meta.SizeBytes != 0 {
		t.Fatalf("meta = %+v (%v)", meta, err)
	}
	got, err := s.GetEntries("x", Query{})
	if err != nil || len(got) != 0 {
		t.Fatalf("got %v (%v)", got, err)
	}
	if p, _ := s.GetPath("x"); p != "" {
		t.Fatalf("path = %q", p)
	}
}

File: packages/provide-uterm-ts/src/recording/recording.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

import { appendFileSync, mkdtempSync, rmSync, statSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { loadGolden } from "../testing/golden.ts";
import {
  InMemoryRecordingStore,
  LocalFileRecordingStore,
  NullRecordingStore,
  normalizeLimit,
  type RecordingEvent,
  type RecordingStore,
} from "./index.ts";

interface RecordingGolden {
  events: RecordingEvent[];
  queries: Array<{
    limit: number;
    offset: number | null;
    event: string | null;
    memory: RecordingEvent[];
    file: RecordingEvent[];
  }>;
  memory_after_end: RecordingEvent[];
  file_after_end: RecordingEvent[];
  memory_meta: { session_id: string; exists: boolean };
  deterministic_meta: { session_id: string; exists: boolean; size_bytes: number };
  memory_meta_missing: { session_id: string; exists: boolean; size_bytes: number };
  file_meta_exists_keys: string[];
  file_meta_exists: { session_id: string; exists: boolean };
  file_meta_missing: { session_id: string; exists: boolean; size_bytes: number; path: string | null };
  file_path_present: boolean;
  file_path_missing: boolean;
  null_meta: { session_id: string; exists: boolean; size_bytes: number };
  null_entries: RecordingEvent[];
  null_path: null;
  limits: Array<{ input: number; normalized: number }>;
}

const golden = loadGolden<RecordingGolden>("recording_golden.json");

let workDir: string;

beforeEach(() => {
  workDir = mkdtempSync(join(tmpdir(), "uterm-recording-"));
});

afterEach(() => {
  rmSync(workDir, { recursive: true, force: true });
});

/** Drop the wall-clock timestamps the lifecycle events carry. */
function stripVolatile(entries: readonly RecordingEvent[]): RecordingEvent[] {
  return entries.map((entry) => {
    if (entry.event === "log_start" || entry.event === "log_stop") {
      const { ts: _ignored, ...rest } = entry;
      return rest;
    }
    return { ...entry };
  });
}

/** Run a store through the lifecycle the corpus recorded. */
async function drive(store: RecordingStore): Promise<void> {
  await store.startSession("s1", { kind: "corpus" });
  await store.appendEvents("s1", golden.events);
}

describe("normalizeLimit", () => {
  it("matches every recorded clamp", () => {
    for (const record of golden.limits) {
      expect({ input: record.input, normalized: normalizeLimit(record.input) }).toStrictEqual(record);
    }
  });

  it("treats zero as the default rather than as empty", () => {
    expect(normalizeLimit(0)).toBe(200);
  });

  it("clamps to the one-to-five-hundred range", () => {
    expect(normalizeLimit(-5)).toBe(1);
    expect(normalizeLimit(99999)).toBe(500);
  });
});

describe("InMemoryRecordingStore", () => {
  it("records an opening event when a session starts", async () => {
    const store = new InMemoryRecordingStore();
    await store.startSession("s1", { kind: "test" });
    const entries = await store.getEntries("s1");
    expect(entries).toHaveLength(1);
    expect(entries[0]).toMatchObject({ event: "log_start", data: { kind: "test" }, session_id: "s1" });
  });

  it("records a closing event when a session ends", async () => {
    const store = new InMemoryRecordingStore();
    await store.startSession("s1", {});
    await store.endSession("s1");
    expect((await store.getEntries("s1")).at(-1)).toMatchObject({ event: "log_stop", session_id: "s1" });
  });

  it("accepts events for a session that was never started", async () => {
    const store = new InMemoryRecordingStore();
    await store.appendEvents("s1", [{ ts: 1, event: "read", data: {} }]);
    expect(await store.getEntries("s1")).toHaveLength(1);
  });

  it("reports an unknown session as absent", async () => {
    expect(await new InMemoryRecordingStore().recordingMeta("nosuch")).toStrictEqual(golden.memory_meta_missing);
  });

  it("has no local path", async () => {
    expect(await new InMemoryRecordingStore().getPath("s1")).toBeNull();
  });

  it("records a closing event for a session that was never started", async () => {
    // There is no session record to mark inactive, but the event still lands
    // so a reader sees a terminated stream rather than one that just stops.
    const store = new InMemoryRecordingStore();
    await store.endSession("s1");
    expect(await store.getEntries("s1")).toHaveLength(1);
  });

  it("returns nothing for a session it has never seen", async () => {
    expect(await new InMemoryRecordingStore().getEntries("nosuch")).toStrictEqual([]);
  });
});

describe("LocalFileRecordingStore", () => {
  it("writes one JSON object per line", async () => {
    const store = new LocalFileRecordingStore(workDir);
    await store.startSession("s1", {});
    await store.appendEvents("s1", [{ ts: 1, event: "read", data: {} }]);
    await store.endSession("s1");
    expect(await store.getEntries("s1")).toHaveLength(3);
  });

  it("reopens the file when events arrive after the session ended", async () => {
    const store = new LocalFileRecordingStore(workDir);
    await store.startSession("s1", {});
    await store.endSession("s1");
    await store.appendEvents("s1", [{ ts: 1, event: "read", data: {} }]);
    expect((await store.getEntries("s1")).at(-1)).toMatchObject({ event: "read" });
  });

  it("appends rather than truncating across store instances", async () => {
    const first = new LocalFileRecordingStore(workDir);
    await first.startSession("s1", {});
    await first.endSession("s1");
    const second = new LocalFileRecordingStore(workDir);
    await second.appendEvents("s1", [{ ts: 1, event: "read", data: {} }]);
    expect(await second.getEntries("s1")).toHaveLength(3);
  });

  it("skips a line that is not valid JSON", async () => {
    // A recording truncated by a crash is still worth serving.
    const store = new LocalFileRecordingStore(workDir);
    await store.appendEvents("s1", [{ ts: 1, event: "read", data: {} }]);
    appendFileSync(join(workDir, "s1.jsonl"), "{not json\n");
    await store.appendEvents("s1", [{ ts: 2, event: "read", data: {} }]);
    expect(await store.getEntries("s1")).toHaveLength(2);
  });

  it("returns nothing for a session with no file", async () => {
    expect(await new LocalFileRecordingStore(workDir).getEntries("nosuch")).toStrictEqual([]);
  });

  it("ends a session it never opened without touching the disk", async () => {
    const store = new LocalFileRecordingStore(workDir);
    await store.endSession("nosuch");
    expect(await store.getPath("nosuch")).toBeNull();
  });

  it("reports metadata for a missing session without inventing a path", async () => {
    expect(await new LocalFileRecordingStore(workDir).recordingMeta("nosuch")).toStrictEqual(golden.file_meta_missing);
  });

  it("reports the on-disk size", async () => {
    const store = new LocalFileRecordingStore(workDir);
    await store.startSession("s1", {});
    const meta = await store.recordingMeta("s1");
    expect(meta.size_bytes).toBe(statSync(join(workDir, "s1.jsonl")).size);
    expect(Object.keys(meta).sort()).toStrictEqual(golden.file_meta_exists_keys);
  });

  it("exposes a path once the file exists and not before", async () => {
    const store = new LocalFileRecordingStore(workDir);
    expect(await store.getPath("s1")).toBeNull();
    await store.startSession("s1", {});
    expect(await store.getPath("s1")).toBe(join(workDir, "s1.jsonl"));
  });

  it("creates the recording directory owner-only", async () => {
    const nested = join(workDir, "nested");
    const store = new LocalFileRecordingStore(nested);
    await store.startSession("s1", {});
    expect(statSync(nested).mode & 0o777).toBe(0o700);
    expect(statSync(join(nested, "s1.jsonl")).mode & 0o777).toBe(0o600);
  });
});

describe("NullRecordingStore", () => {
  it("discards every write and returns nothing", async () => {
    const store = new NullRecordingStore();
    await store.startSession("s1", { kind: "test" });
    await store.appendEvents("s1", golden.events);
    await store.endSession("s1");
    expect(await store.getEntries("s1")).toStrictEqual(golden.null_entries);
    expect(await store.recordingMeta("s1")).toStrictEqual(golden.null_meta);
    expect(await store.getPath("s1")).toBe(golden.null_path);
  });
});

describe("differential parity with CPython", () => {
  it("matches every recorded query on both stores", async () => {
    const memory = new InMemoryRecordingStore();
    const file = new LocalFileRecordingStore(workDir);
    await drive(memory);
    await drive(file);

    for (const record of golden.queries) {
      const options = { limit: record.limit, offset: record.offset, event: record.event };
      expect({
        query: [record.limit, record.offset, record.event],
        memory: stripVolatile(await memory.getEntries("s1", options)),
        file: stripVolatile(await file.getEntries("s1", options)),
      }).toStrictEqual({
        query: [record.limit, record.offset, record.event],
        memory: record.memory,
        file: record.file,
      });
    }
    expect(golden.queries.length).toBeGreaterThan(10);
  });

  it("matches the recorded event stream after each store is closed", async () => {
    const memory = new InMemoryRecordingStore();
    await drive(memory);
    await memory.endSession("s1");
    expect(stripVolatile(await memory.getEntries("s1", { limit: 200 }))).toStrictEqual(golden.memory_after_end);

    const file = new LocalFileRecordingStore(workDir);
    await drive(file);
    await file.endSession("s1");
    expect(stripVolatile(await file.getEntries("s1", { limit: 200 }))).toStrictEqual(golden.file_after_end);
  });

  it("matches the recorded in-memory size, measured with CPython separators", async () => {
    // Measured over the fixed events only. A store that also holds a
    // log_start carries a wall-clock timestamp whose rendered length varies
    // between runs, so its size is not a reproducible thing to assert.
    const store = new InMemoryRecordingStore();
    await store.appendEvents("s1", golden.events);
    expect(await store.recordingMeta("s1")).toStrictEqual(golden.deterministic_meta);
  });

  it("reports a started session as existing", async () => {
    const memory = new InMemoryRecordingStore();
    await drive(memory);
    const meta = await memory.recordingMeta("s1");
    expect({ session_id: meta.session_id, exists: meta.exists }).toStrictEqual(golden.memory_meta);
  });
});
